### ISSUE TITLE
Identifier conflict in Debug derive implementation

### ISSUE THREAD

**Issue by AnthonyMikh** [rust-lang/rust#60055]
TL;DR: sophisticated user code can make deriving `Debug` impossible.

I've written the code which looks [like this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=52c7774bfd191feca480297de8247a47) in reduced form:

```rust
struct X;

trait FieldOf<T: ?Sized> {
    type ValueType;
    fn of(val: &T) -> &Self::ValueType;
}

trait GetFieldExt {
    fn field<F: FieldOf<Self>>(&self) -> &F::ValueType;
}

impl<T> GetFieldExt for T {
    fn field<F: FieldOf<Self>>(&self) -> &F::ValueType {
        F::of(self)
    }
}

//#[derive(Debug)] <-- uncomment to trigger error
struct Vec3(i32, i32, i32);

impl FieldOf<Vec3> for X {
    type ValueType = i32;
    fn of(val: &Vec3) -> &Self::ValueType {
        &val.0
    }
}

fn main() {
    let vec = Vec3(0, 0, 0);
    println!("{:?}", vec.field::<X>());
}
```

I compiles and works just fine. However, trying to add `#[derive(Debug)]` to `Vec3` makes this code fail to compile with the following error:

```
error[E0061]: this function takes 0 parameters but 1 parameter was supplied
  --> src/main.rs:18:10
   |
9  |     fn field<F: FieldOf<Self>>(&self) -> &F::ValueType;
   |     --------------------------------------------------- defined here
...
18 | #[derive(Debug)]
   |          ^^^^^ expected 0 parameters

error: aborting due to previous error
```

## Meta

I've used playground, so it is rustc-1.34.0-stable, the latest at the moment.

As one of my friends @mersinvald pointed out, the possible cause may lie [here](https://github.com/rust-lang/rust/blob/c033c6e47c1a84e2902c75a1b5ec161362f34f18/src/libsyntax_ext/deriving/debug.rs#L118), in implementation of derive, where method is called without fully qualified syntax.

**Comment by AnthonyMikh** [comment]
I've managed to reduce the [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c96c5bd4b8f10f4a0eded1e275cccf14):

```rust
trait Field {
    fn field(&self) {}
}

impl<T> Field for T {}

#[derive(Debug)]
struct Foo(());

fn main() {
    Foo(()).field();
}
```

It fails to compile with similar message:

```
error[E0061]: this function takes 0 parameters but 1 parameter was supplied
 --> src/main.rs:7:10
  |
2 |     fn field(&self) {}
  |     --------------- defined here
...
7 | #[derive(Debug)]
  |          ^^^^^ expected 0 parameters

```

It looks like in order to trigger the error one should supply a type with fields with a method named `field` with non-zero number of parameters via blanket `impl`. Implementing an inherent method or implementing the trait directly doesn't trigger this error. Error is not triggered by fieldless structs and enums with fieldless variants.

**Comment by AnthonyMikh** [comment]
Looks like it was fixed in Rust 1.39, this version as well as newer ones compile the example just fine.
